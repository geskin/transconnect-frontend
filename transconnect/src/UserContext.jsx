import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

