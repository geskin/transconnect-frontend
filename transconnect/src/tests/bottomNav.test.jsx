import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import CommentsBottomNavigation from "../utils/bottomNav";
import UserContext from "../UserContext";
import TransconnectApi from "../api";

vi.mock("../api", () => ({
    default: {
        getComments: vi.fn(),
        createComment: vi.fn(),
        deleteComment: vi.fn(),
        editComment: vi.fn(),
        getComment: vi.fn()
    }
}));

describe("CommentsBottomNavigation Component", () => {
    const mockUser = {
        id: 1,
        username: "testuser",
        role: "USER"
    };
    const mockComments = [
        { id: 1, content: "First comment", createdAt: "2024-03-14T12:00:00Z", author: { username: "testuser" } },
        { id: 2, content: "Second comment", createdAt: "2024-03-14T13:00:00Z", author: { username: "anotheruser" } }
    ];

    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("renders without crashing", () => {
        render(
            <UserContext.Provider value={{ currUser: mockUser }}>
                <CommentsBottomNavigation postId={123} />
            </UserContext.Provider>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <UserContext.Provider value={{ currUser: mockUser }}>
                <CommentsBottomNavigation postId={123} />
            </UserContext.Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("fetches and displays comments", async () => {
        TransconnectApi.getComments.mockResolvedValue(mockComments);

        render(
            <UserContext.Provider value={{ currUser: mockUser }}>
                <CommentsBottomNavigation postId={123} />
            </UserContext.Provider>
        );

        await waitFor(() => {
            expect(TransconnectApi.getComments).toHaveBeenCalledWith(123);
        });
    });

    // it("adds a comment when submitted", async () => {
    //     TransconnectApi.getComments.mockResolvedValue([]);
    //     TransconnectApi.createComment.mockResolvedValue({
    //         id: 3,
    //         content: "New comment",
    //         createdAt: "2024-03-14T14:00:00Z",
    //         author: { username: "testuser" }
    //     });

    //     render(
    //         <UserContext.Provider value={{ currUser: mockUser }}>
    //             <CommentsBottomNavigation postId={123} />
    //         </UserContext.Provider>
    //     );

    //     fireEvent.click(screen.getByText("Comment"));
    //     fireEvent.change(screen.getByText("Add a comment..."), { target: { value: "New comment" } });
    //     fireEvent.click(screen.getByText("Submit"));

    //     await waitFor(() => {
    //         expect(TransconnectApi.createComment).toHaveBeenCalledWith(123, "New comment", mockUser.id);
    //     });
    // });

    // it("deletes a comment when the delete button is clicked", async () => {
    //     TransconnectApi.getComments.mockResolvedValue(mockComments);
    //     TransconnectApi.getComment.mockResolvedValue(mockComments[0]);
    //     TransconnectApi.deleteComment.mockResolvedValue(true);

    //     render(
    //         <UserContext.Provider value={{ currUser: mockUser }}>
    //             <CommentsBottomNavigation postId={123} />
    //         </UserContext.Provider>
    //     );

    //     await waitFor(() => expect(screen.getByText("First comment")).toBeInTheDocument());

    //     fireEvent.click(screen.getAllByLabelText("delete")[0]);

    //     await waitFor(() => {
    //         expect(TransconnectApi.deleteComment).toHaveBeenCalledWith(123, 1, mockComments[0], "testuser");
    //         expect(screen.queryByText("First comment")).not.toBeInTheDocument();
    //     });
    // });

    // it("allows a user to edit their own comment", async () => {
    //     TransconnectApi.getComments.mockResolvedValue(mockComments);
    //     TransconnectApi.editComment.mockResolvedValue({ content: "Updated comment" });

    //     render(
    //         <UserContext.Provider value={{ currUser: mockUser }}>
    //             <CommentsBottomNavigation postId={123} />
    //         </UserContext.Provider>
    //     );

    //     fireEvent.click(screen.getAllByLabelText("edit")[0]);

    //     fireEvent.change(screen.getByDisplayValue("First comment"), { target: { value: "Updated comment" } });

    //     fireEvent.click(screen.getByLabelText("save"));

    //     await waitFor(() => {
    //         expect(TransconnectApi.editComment).toHaveBeenCalledWith(123, 1, "Updated comment", "testuser");
    //         expect(screen.getByText("Updated comment")).toBeInTheDocument();
    //     });
    // });
});
