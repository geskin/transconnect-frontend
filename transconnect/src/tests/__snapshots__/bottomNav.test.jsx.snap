// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentsBottomNavigation Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1nl0624"
  >
    <ul
      class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
    >
      No comments yet!
    </ul>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-1ebsret-MuiPaper-root"
      style="--Paper-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBottomNavigation-root css-7ax0nd-MuiBottomNavigation-root"
      >
        <button
          class="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected css-1scjuq2-MuiButtonBase-root-MuiBottomNavigationAction-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="AddCommentIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4zM17 11h-4v4h-2v-4H7V9h4V5h2v4h4z"
            />
          </svg>
          <span
            class="MuiBottomNavigationAction-label Mui-selected css-1kcvi62-MuiBottomNavigationAction-label"
          >
            Comment
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
